File: conformance/java/src/main/java/com/birdayz/conformance/ConformanceTest.java
package com.birdayz.conformance;

import com.apple.foundationdb.record.RecordMetaData;
import com.apple.foundationdb.record.RecordMetaDataBuilder;
import com.apple.foundationdb.record.provider.foundationdb.FDBDatabase;
import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseFactory;
import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore;
import com.apple.foundationdb.record.provider.foundationdb.FDBStoredRecord;
import com.apple.foundationdb.record.metadata.Key;
import com.apple.foundationdb.subspace.Subspace;
import com.apple.foundationdb.tuple.Tuple;
import com.google.protobuf.Message;

/**
 * Java conformance test for FoundationDB Record Layer.
 * This class can read data written by the Go implementation and write data for Go to read.
 */
public class ConformanceTest {
    private static final String SUBSPACE_NAME = "conformance_test";
    private static final FDBDatabase db = FDBDatabaseFactory.instance().getDatabase();
    
    public static void main(String[] args) {
        if (args.length == 0) {
            System.err.println("Usage: ConformanceTest <read|write>");
            System.exit(1);
        }
        
        String mode = args[0];
        
        try {
            if ("read".equals(mode)) {
                readData();
            } else if ("write".equals(mode)) {
                writeData();
            } else {
                System.err.println("Invalid mode: " + mode + ". Use 'read' or 'write'");
                System.exit(1);
            }
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
    
    private static void readData() {
        System.out.println("=== JAVA READ DEBUG ===");
        // Create metadata - must match Go implementation exactly
        RecordMetaDataBuilder metaDataBuilder = RecordMetaData.newBuilder()
                .setRecords(RecordLayerDemoProto.getDescriptor());
        
        metaDataBuilder.getRecordType("Order")
                .setPrimaryKey(Key.Expressions.field("order_id"));
        
        RecordMetaData recordMetaData = metaDataBuilder.build();
        
        // Create subspace - must match Go implementation
        Subspace subspace = new Subspace(Tuple.from(SUBSPACE_NAME));
        System.out.println("Java subspace bytes: " + java.util.Arrays.toString(subspace.pack()));
        
        // Read the data Go wrote
        FDBStoredRecord<Message> storedRecord = db.run(context -> {
            FDBRecordStore recordStore = FDBRecordStore.newBuilder()
                    .setMetaDataProvider(recordMetaData)
                    .setContext(context)
                    .setSubspace(subspace)
                    .open();
            
            // Try to load order ID 1001 (what Go writes)
            Tuple primaryKey = Tuple.from(1001L);
            System.out.println("Java primary key: " + primaryKey);
            System.out.println("Java looking for record with this key structure");
            return recordStore.loadRecord(primaryKey);
        });
        
        if (storedRecord == null) {
            System.out.println("ERROR: No record found with order_id 1001");
            return;
        }
        
        // Parse the protobuf record
        RecordLayerDemoProto.Order order = RecordLayerDemoProto.Order.newBuilder()
                .mergeFrom(storedRecord.getRecord())
                .build();
        
        // Verify the data matches what Go wrote
        if (order.getOrderId() == 1001 && 
            order.getPrice() == 25 && 
            order.getFlower().getType().equals("Rose") &&
            order.getFlower().getColor() == RecordLayerDemoProto.Color.RED) {
            
            System.out.println("SUCCESS: Found order " + order.getOrderId() + 
                             " with price " + order.getPrice() + 
                             " and flower " + order.getFlower().getType() + 
                             " (" + order.getFlower().getColor() + ")");
        } else {
            System.out.println("ERROR: Data mismatch. Got: " + order);
        }
    }
    
    private static void writeData() {
        // Create metadata - must match Go implementation exactly
        RecordMetaDataBuilder metaDataBuilder = RecordMetaData.newBuilder()
                .setRecords(RecordLayerDemoProto.getDescriptor());
        
        metaDataBuilder.getRecordType("Order")
                .setPrimaryKey(Key.Expressions.field("order_id"));
        
        RecordMetaData recordMetaData = metaDataBuilder.build();
        
        // Create subspace - must match Go implementation
        Subspace subspace = new Subspace(Tuple.from(SUBSPACE_NAME));
        
        // Write test data for Go to read
        db.run(context -> {
            FDBRecordStore recordStore = FDBRecordStore.newBuilder()
                    .setMetaDataProvider(recordMetaData)
                    .setContext(context)
                    .setSubspace(subspace)
                    .createOrOpen();
            
            // Create test order - Go will try to read this (order ID 2002)
            RecordLayerDemoProto.Order order = RecordLayerDemoProto.Order.newBuilder()
                    .setOrderId(2002)
                    .setPrice(50)
                    .setFlower(RecordLayerDemoProto.Flower.newBuilder()
                            .setType("Tulip")
                            .setColor(RecordLayerDemoProto.Color.BLUE))
                    .build();
            
            recordStore.saveRecord(order);
            System.out.println("Java wrote order: " + order.getOrderId());
            return null;
        });
    }
}

File: conformance/conformance_test.go
package conformance

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"

	"github.com/apple/foundationdb/bindings/go/src/fdb"
	"github.com/apple/foundationdb/bindings/go/src/fdb/subspace"
	"github.com/apple/foundationdb/bindings/go/src/fdb/tuple"

	"github.com/birdayz/fdb-record-layer-go/gen"
	"github.com/birdayz/fdb-record-layer-go/pkg/recordlayer"
	"google.golang.org/protobuf/proto"
)

const (
	// Test subspace name - must match Java test
	testSubspaceName = "conformance_test"
)

func TestGoWriteJavaRead(t *testing.T) {
	// Skip if FDB is not available
	if os.Getenv("SKIP_FDB_TESTS") != "" {
		t.Skip("Skipping FDB conformance tests (SKIP_FDB_TESTS set)")
	}

	// Initialize FDB
	fdb.MustAPIVersion(720)
	db := fdb.MustOpenDefault()
	recordDB := recordlayer.NewFDBDatabase(db)

	// Setup metadata - same as our getting started example
	fileDesc := gen.File_record_layer_demo_proto
	metaDataBuilder := recordlayer.NewRecordMetaDataBuilder().SetRecords(fileDesc)
	metaDataBuilder.GetRecordType("Order").SetPrimaryKey(recordlayer.Field("order_id"))
	recordMetaData := metaDataBuilder.Build()

	// Create test subspace - must match Java's new Subspace(Tuple.from(SUBSPACE_NAME))
	keyspace := subspace.FromBytes(tuple.Tuple{testSubspaceName}.Pack())

	// Write test data with Go implementation
	err := writeTestDataWithGo(recordDB, recordMetaData, keyspace)
	if err != nil {
		t.Fatalf("Failed to write test data with Go: %v", err)
	}

	// Now call Java to read the data
	javaOutput, err := callJavaReader()
	if err != nil {
		t.Fatalf("Failed to call Java reader: %v", err)
	}

	// Verify Java could read the data correctly
	expectedOutput := "SUCCESS: Found order 1001 with price 25 and flower Rose (RED)"
	// Extract just the meaningful line from the Gradle output
	lines := strings.Split(javaOutput, "\n")
	var meaningfulOutput string
	for _, line := range lines {
		if strings.HasPrefix(line, "SUCCESS:") || strings.HasPrefix(line, "ERROR:") {
			meaningfulOutput = line
			break
		}
	}
	
	if meaningfulOutput != expectedOutput {
		t.Fatalf("Java reader output mismatch.\nExpected: %s\nGot: %s\nFull output: %s", expectedOutput, meaningfulOutput, javaOutput)
	}

	t.Logf("Conformance test passed! Java successfully read Go-written data: %s", javaOutput)
}

func TestJavaWriteGoRead(t *testing.T) {
	// Skip if FDB is not available
	if os.Getenv("SKIP_FDB_TESTS") != "" {
		t.Skip("Skipping FDB conformance tests (SKIP_FDB_TESTS set)")
	}

	// Call Java to write test data
	err := callJavaWriter()
	if err != nil {
		t.Fatalf("Failed to call Java writer: %v", err)
	}

	// Initialize FDB
	fdb.MustAPIVersion(720)
	db := fdb.MustOpenDefault()
	recordDB := recordlayer.NewFDBDatabase(db)

	// Setup metadata
	fileDesc := gen.File_record_layer_demo_proto
	metaDataBuilder := recordlayer.NewRecordMetaDataBuilder().SetRecords(fileDesc)
	metaDataBuilder.GetRecordType("Order").SetPrimaryKey(recordlayer.Field("order_id"))
	recordMetaData := metaDataBuilder.Build()

	// Create test subspace - must match Java's new Subspace(Tuple.from(SUBSPACE_NAME))
	keyspace := subspace.FromBytes(tuple.Tuple{testSubspaceName}.Pack())

	// Read test data with Go implementation
	orderData, err := readTestDataWithGo(recordDB, recordMetaData, keyspace)
	if err != nil {
		t.Fatalf("Failed to read test data with Go: %v", err)
	}

	// Verify the data matches what Java wrote
	if orderData.OrderId == nil || *orderData.OrderId != 2002 {
		t.Fatalf("Expected order ID 2002, got %v", orderData.OrderId)
	}
	if orderData.Price == nil || *orderData.Price != 50 {
		t.Fatalf("Expected price 50, got %v", orderData.Price)
	}
	if orderData.Flower == nil || orderData.Flower.Type == nil || *orderData.Flower.Type != "Tulip" {
		t.Fatalf("Expected flower type 'Tulip', got %v", orderData.Flower)
	}

	t.Logf("Conformance test passed! Go successfully read Java-written data: order_id=%d, price=%d, flower=%s",
		*orderData.OrderId, *orderData.Price, *orderData.Flower.Type)
}

func writeTestDataWithGo(recordDB *recordlayer.FDBDatabase, metaData *recordlayer.RecordMetaData, keyspace subspace.Subspace) error {
	_, err := recordDB.Run(context.Background(), func(ctx *recordlayer.FDBRecordContext) (interface{}, error) {
		store, err := recordlayer.NewStoreBuilder().
			SetContext(ctx).
			SetMetaDataProvider(metaData).
			SetSubspace(keyspace).
			CreateOrOpen()
		if err != nil {
			return nil, err
		}

		// Create test order - same structure Java expects
		order := &gen.Order{
			OrderId: proto.Int64(1001),
			Price:   proto.Int32(25),
			Flower: &gen.Flower{
				Type:  proto.String("Rose"),
				Color: gen.Color_RED.Enum(),
			},
		}

		_, err = store.SaveRecord(order)
		return nil, err
	})

	return err
}

func readTestDataWithGo(recordDB *recordlayer.FDBDatabase, metaData *recordlayer.RecordMetaData, keyspace subspace.Subspace) (*gen.Order, error) {
	result, err := recordDB.Run(context.Background(), func(ctx *recordlayer.FDBRecordContext) (interface{}, error) {
		store, err := recordlayer.NewStoreBuilder().
			SetContext(ctx).
			SetMetaDataProvider(metaData).
			SetSubspace(keyspace).
			CreateOrOpen()
		if err != nil {
			return nil, err
		}

		// Try to load the record Java wrote (order ID 2002)
		primaryKey := tuple.Tuple{int64(2002)}
		storedRecord, err := store.LoadRecord(primaryKey)
		if err != nil {
			return nil, err
		}

		if storedRecord == nil {
			return nil, nil
		}

		// Extract the actual deserialized Order from the stored record
		order, ok := storedRecord.Record.(*gen.Order)
		if !ok {
			return nil, fmt.Errorf("expected *gen.Order, got %T", storedRecord.Record)
		}

		return order, nil
	})

	if err != nil {
		return nil, err
	}

	if result == nil {
		return nil, nil
	}

	return result.(*gen.Order), nil
}

func callJavaReader() (string, error) {
	javaDir := filepath.Join(".", "java")
	cmd := exec.Command("./gradlew", "run", "--args=read")
	cmd.Dir = javaDir

	output, err := cmd.CombinedOutput()
	if err != nil {
		return string(output), err // Return output even on error for debugging
	}

	return string(output), nil
}

func callJavaWriter() error {
	javaDir := filepath.Join(".", "java")
	cmd := exec.Command("./gradlew", "run", "--args=write")
	cmd.Dir = javaDir

	return cmd.Run()
}